index.js

```diff
@@ -180,2669 +180,2588 @@
-  const Date = NullaryTypeWithUrl
-    ('Date')
-    ([])
-    (x => type (x) === 'Date');
-
-  //# ValidDate :: Type
-  //.
-  //. Type comprising every [`Date`][] value except `new Date (NaN)`.
-  const ValidDate = NullaryTypeWithUrl
-    ('ValidDate')
-    ([Date])
-    (date => !(globalThis.Number.isNaN (date.valueOf ())));
-
-  //# Descending :: Type -> Type
-  //.
-  //. [Descending][] type constructor.
-  const Descending = UnaryTypeWithUrl
-    ('Descending')
-    ([])
-    (x => type (x) === 'sanctuary-descending/Descending@1')
-    (descending => descending);
-
-  //# Either :: Type -> Type -> Type
-  //.
-  //. [Either][] type constructor.
-  const Either = BinaryTypeWithUrl
-    ('Either')
-    ([])
-    (x => type (x) === 'sanctuary-either/Either@1')
-    (either => either.isLeft ? [either.value] : [])
-    (either => either.isLeft ? [] : [either.value]);
-
-  //# Error :: Type
-  //.
-  //. Type comprising every Error value, including values of more specific
-  //. constructors such as [`SyntaxError`][] and [`TypeError`][].
-  const Error = NullaryTypeWithUrl
-    ('Error')
-    ([])
-    (x => type (x) === 'Error');
-
-  //# Fn :: Type -> Type -> Type
-  //.
-  //. Binary type constructor for unary function types. `$.Fn (I) (O)`
-  //. represents `I -> O`, the type of functions that take a value of
-  //. type `I` and return a value of type `O`.
-  const Fn = $1 => $2 => Function ([$1, $2]);
-
-  //# Function :: NonEmpty (Array Type) -> Type
-  //.
-  //. Constructor for Function types.
-  //.
-  //. Examples:
-  //.
-  //.   - `$.Function ([$.Date, $.String])` represents the `Date -> String`
-  //.     type; and
-  //.   - `$.Function ([a, b, a])` represents the `(a, b) -> a` type.
-  const Function = types => (
-    _Type (
-      FUNCTION,
-      '',
-      '',
-      types.length,
-      (outer, inner) => {
-        const repr = (
-          types
-          .slice (0, -1)
-          .map ((t, idx) =>
-            t.type === FUNCTION
-            ? outer ('(') + inner (`$${idx + 1}`) (show (t)) + outer (')')
-            : inner (`$${idx + 1}`) (show (t))
-          )
-          .join (outer (', '))
-        );
-        return (
-          (types.length === 2 ? repr : outer ('(') + repr + outer (')')) +
-          outer (' -> ') +
-          inner (`$${types.length}`)
-                (show (types[types.length - 1]))
-        );
-      },
-      [AnyFunction],
-      env => x => true,
-      types.map ((t, idx) => [`$${idx + 1}`, x => [], t])
-    )
-  );
+//# NonEmpty :: Type -> Type
+//.
+//. Constructor for non-empty types. `$.NonEmpty ($.String)`, for example, is
+//. the type comprising every [`String`][] value except `''`.
+//.
+//. The given type must satisfy the [Monoid][] and [Setoid][] specifications.
+export const NonEmpty = UnaryTypeWithUrl
+  ('NonEmpty')
+  ([])
+  (x => Z.Monoid.test (x) &&
+        Z.Setoid.test (x) &&
+        !(Z.equals (x, Z.empty (x.constructor))))
+  (monoid => [monoid]);
```

Comment:
I moved this block near the top of the section as other type definitions depend on `NonEmpty`.